File: trees/deepest_leaves_sum.cpp
//leetcode problem - 1302 : deepest leaves sum


class Solution {
public:
    int depth=0, ans=0;
    void solve(TreeNode* root, int k){
        if(!root) return;
        if(root->left==NULL and root->right==NULL)
            depth=max(depth,k);
        solve(root->left,k+1);
        solve(root->right, k+1);
    }
    void findans(TreeNode* root, int k){
        if(!root) return;
        if(k==depth)
            ans+=root->val;
        findans(root->left, k+1);
        findans(root->right, k+1);
    }
    int deepestLeavesSum(TreeNode* root) {
        solve(root,1);
        findans(root,1);
        return ans;
    }
};

File: trees/deepest_leaves_sum.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def deepest_leaves_sum(self, root):
        def dfs1(root):
            if not root: return 0
            return max(dfs1(root.left), dfs1(root.right))+1
        def dfs2(root, dep):
            if not root: return
            if dep == max_depth: self.total += root.val
            dfs2(root.left, dep+1)
            dfs2(root.right, dep+1)
        self.total = 0
        max_depth = dfs1(root)
        dfs2(root, 1)
        return self.total

